### PR TITLE
Persist index changes into CSV file

### PR DESCRIPTION
**Pull request by berezkin88**

### CODE REVIEW

File: xs2a-adapter-aspsp-registry/src/main/java/de/adorsys/xs2a/adapter/registry/AspspCsvServiceImpl.java [hunk 1]
```diff
@@ -16,12 +16,22 @@
...
+import java.nio.file.Files;
+import java.nio.file.Paths;
 import java.util.List;
 import java.util.stream.Collectors;
 
 public class AspspCsvServiceImpl implements AspspCsvService {
 
+    private static final String SLASH = File.separator;
+    private static final String CSV_ASPSP_ADAPTER_CONFIG_FILE_PATH_PROPERTY = "csv.aspsp.adapter.config.file.path";
+    private static final String DEFAULT_CSV_ASPSP_ADAPTER_CONFIG_FILE_PATH = System.getProperty("user.dir") + SLASH
```
Comment (spetrychenko) [inline xs2a-adapter-aspsp-registry/src/main/java/de/adorsys/xs2a/adapter/registry/AspspCsvServiceImpl.java]: should be rewritten

File: xs2a-adapter-aspsp-registry/src/main/java/de/adorsys/xs2a/adapter/registry/AspspCsvServiceImpl.java [hunk 2]
```diff
@@ -56,6 +66,22 @@ public void importCsv(byte[] bytes) {
         aspspRepository.saveAll(aspsps);
     }
 
+    @Override
+    public void rewriteOriginalCsv() {
```
Comment (spetrychenko) [inline xs2a-adapter-aspsp-registry/src/main/java/de/adorsys/xs2a/adapter/registry/AspspCsvServiceImpl.java]: should be renamed

File: xs2a-adapter-aspsp-registry/src/main/java/de/adorsys/xs2a/adapter/registry/AspspCsvServiceImpl.java [hunk 3]
```diff
@@ -65,11 +91,11 @@ private String toCsvString(AspspCsvRecord aspsp) {
             String columnName = "aspspName";
             int nameColumnIndex = schema.column(columnName).getIndex();
             schema = mapper
-                .configure(CsvGenerator.Feature.STRICT_CHECK_FOR_QUOTING, true)
-                .schemaFor(AspspCsvRecord.class)
-                .rebuild()
-                .replaceColumn(nameColumnIndex, new CsvSchema.Column(nameColumnIndex, columnName))
-                .build();
+                         .configure(CsvGenerator.Feature.STRICT_CHECK_FOR_QUOTING, true)
```
Comment (spetrychenko) [inline xs2a-adapter-aspsp-registry/src/main/java/de/adorsys/xs2a/adapter/registry/AspspCsvServiceImpl.java]: rollback all tab spaces in the class

File: xs2a-adapter-service-api/src/main/java/de/adorsys/xs2a/adapter/service/AspspCsvService.java [hunk 4]
```diff
@@ -47,4 +46,20 @@
...
+     * created via Registry UI (manually or by importing a new CSV). The original CSV
+     * file is searched under "csv.aspsp.adapter.config.file.path" property, if the property
+     * is not specified, the default file path is taken into consideration.
+     * <p>
+     * {@link java.nio.file.Files} and {@link java.nio.file.Paths} are used for
+     * re-writing the CSV
+     *
+     * @throws RegistryIOException if writing a new CSV fails
+     */
+    void rewriteOriginalCsv();
```
Comment (spetrychenko) [inline xs2a-adapter-service-api/src/main/java/de/adorsys/xs2a/adapter/service/AspspCsvService.java]: rename please

File: xs2a-adapter-aspsp-registry/src/main/java/de/adorsys/xs2a/adapter/registry/AspspCsvServiceImpl.java [hunk 5]
```diff
@@ -56,6 +62,22 @@ public void importCsv(byte[] bytes) {
         aspspRepository.saveAll(aspsps);
     }
 
+    @Override
+    public void saveCsv() {
+        try {
+            if (CSV_ASPSP_ADAPTER_CONFIG_FILE_PATH == null || CSV_ASPSP_ADAPTER_CONFIG_FILE_PATH.isEmpty()) {
+//                String filePath = getClass().getClassLoader().getResource(DEFAULT_CSV_ASPSP_ADAPTER_CONFIG_FILE).getPath();
```
Comment (spetrychenko) [inline xs2a-adapter-aspsp-registry/src/main/java/de/adorsys/xs2a/adapter/registry/AspspCsvServiceImpl.java]: remove commented lines

File: xs2a-adapter-aspsp-registry/src/main/java/de/adorsys/xs2a/adapter/registry/AspspCsvServiceImpl.java [hunk 6]
```diff
@@ -16,15 +16,21 @@
...
-import java.io.IOException;
+import java.io.*;
+import java.nio.file.Files;
+import java.nio.file.Paths;
 import java.util.List;
 import java.util.stream.Collectors;
 
 public class AspspCsvServiceImpl implements AspspCsvService {
 
+    private static final String CSV_ASPSP_ADAPTER_CONFIG_FILE_PATH = System.getenv("csv.aspsp.adapter.config.file.path");
```
Comment (spetrychenko) [inline xs2a-adapter-aspsp-registry/src/main/java/de/adorsys/xs2a/adapter/registry/AspspCsvServiceImpl.java]: replace by `PropertyUtil` usage

File: xs2a-adapter-aspsp-registry/src/main/java/de/adorsys/xs2a/adapter/registry/AspspCsvServiceImpl.java [hunk 7]
```diff
@@ -56,6 +62,22 @@ public void importCsv(byte[] bytes) {
         aspspRepository.saveAll(aspsps);
     }
 
+    @Override
+    public void saveCsv() {
+        try {
+            if (CSV_ASPSP_ADAPTER_CONFIG_FILE_PATH == null || CSV_ASPSP_ADAPTER_CONFIG_FILE_PATH.isEmpty()) {
+//                String filePath = getClass().getClassLoader().getResource(DEFAULT_CSV_ASPSP_ADAPTER_CONFIG_FILE).getPath();
+//                Files.write(Paths.get(filePath), exportCsv());
+                throw new RegistryIOException("Adapter config file path is not set or Configuration File does not exist");
```
Comment (artemsnisarenko) [inline xs2a-adapter-aspsp-registry/src/main/java/de/adorsys/xs2a/adapter/registry/AspspCsvServiceImpl.java]: "Adapter config file" to me refers to `adapter.config.properties`. Also what "Configuration File" may not exist here and how did you determine that? It's not an IOException.

File: xs2a-adapter-aspsp-registry/src/main/java/de/adorsys/xs2a/adapter/registry/AspspCsvServiceImpl.java [hunk 8]
```diff
@@ -56,6 +62,22 @@ public void importCsv(byte[] bytes) {
...
+            } else {
+                Files.write(Paths.get(CSV_ASPSP_ADAPTER_CONFIG_FILE_PATH), exportCsv());
+            }
+        } catch (IOException e) {
+            log.error("Exception occurred while re-writing aspsps into the CSV: {}", e.getMessage());
```
Comment (artemsnisarenko) [inline xs2a-adapter-aspsp-registry/src/main/java/de/adorsys/xs2a/adapter/registry/AspspCsvServiceImpl.java]: Don't log and rethrow an exception. https://stackoverflow.com/questions/6639963/why-is-log-and-throw-considered-an-anti-pattern

File: xs2a-adapter-aspsp-registry/src/main/java/de/adorsys/xs2a/adapter/registry/exception/RegistryIOException.java [hunk 9]
```diff
@@ -8,4 +8,8 @@
     public RegistryIOException(IOException cause) {
         super(Objects.requireNonNull(cause));
     }
+
+    public RegistryIOException(String s) {
```
Comment (artemsnisarenko) [inline xs2a-adapter-aspsp-registry/src/main/java/de/adorsys/xs2a/adapter/registry/exception/RegistryIOException.java]: Please remove. See  comment on `saveCsv`.

File: xs2a-adapter-aspsp-registry/src/test/java/de/adorsys/xs2a/adapter/registry/AspspCsvServiceImplTest.java [hunk 10]
```diff
@@ -161,6 +165,26 @@ public void importCsv_oldSchema() {
         assertThat(output.getId()).isEqualTo(ASPSP.getId());
     }
 
+    @Test
+    public void persist() throws IOException {
```
Comment (artemsnisarenko) [inline xs2a-adapter-aspsp-registry/src/test/java/de/adorsys/xs2a/adapter/registry/AspspCsvServiceImplTest.java]: What is this testing?

**Comment by berezkin88** [comment]
all remarks are managed

**Pull request merged**